### PR TITLE
Fix logprobs and logits-related OOM

### PR DESCRIPTION
**Pull request by pjin-nvidia**

### CODE REVIEW

File: nemo_rl/distributed/model_utils.py [hunk 1]
```diff
@@ -149,6 +265,7 @@ def dtensor_from_parallel_logits_to_logprobs(
     tp_group: torch.distributed.ProcessGroup,
     inference_only: bool = False,
     seq_index: Optional[torch.Tensor] = None,
+    chunk_size: Optional[int] = None,
```
Comment (wangshangsam) [inline nemo_rl/distributed/model_utils.py:266]: Update the docstring below, please; i.e.,
```python
    """
        ...
        seq_index (Optional[torch.Tensor]): Sequence index tensor with shape [seq_len].
            It is only provided for cp sharded logits. It represents how tensor is sharded across the sequence dimension.
        chunk_size (Optional[int]): ...
    """
```

File: nemo_rl/distributed/model_utils.py [hunk 2]
```diff
@@ -221,6 +352,7 @@ def from_parallel_logits_to_logprobs(
     tp_group: torch.distributed.ProcessGroup,
     inference_only: bool = False,
     cp_group: Optional[torch.distributed.ProcessGroup] = None,
+    chunk_size: Optional[int] = None,
```
Comment (wangshangsam) [inline nemo_rl/distributed/model_utils.py:351]: Please update the docstring below.

File: nemo_rl/models/dtensor/parallelize.py [hunk 3]
```diff
@@ -620,6 +620,7 @@ def get_logprobs_from_vocab_parallel_logits(
     vocab_parallel_logits: DTensor,
     input_ids: torch.Tensor | DTensor,
     seq_index: Optional[torch.Tensor] = None,
+    chunk_size: Optional[int] = None,
```
Comment (wangshangsam) [inline nemo_rl/models/dtensor/parallelize.py:699]: Please update the docstring below.

File: nemo_rl/tron/model.py [hunk 4]
```diff
@@ -0,0 +1,182 @@
+# Copyright (c) 2025, NVIDIA CORPORATION.  All rights reserved.
```
Comment (wangshangsam) [inline nemo_rl/tron/model.py]: I'm wondering whether it would make more sense to put it in the actual nemo instead of here? @terrykong

File: nemo_rl/tron/model.py [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+    TorchFullyShardedDataParallel,
+)
+from megatron.core.enums import ModelType
+from megatron.core.fp8_utils import is_float8tensor
+
+from nemo.collections.llm.gpt.model.base import GPTConfig
+from nemo.collections.llm.t5.model.t5 import T5Config
+
+
+def get_model_from_config_no_float32(
```
Comment (terrykong) [inline nemo_rl/tron/model.py]: was this function copied from somewhere? if so, what changes were made?

File: nemo_rl/distributed/model_utils.py [hunk 6]
```diff
@@ -141,6 +143,123 @@ def backward(
         return grad_input, None, None, None, None, None, None
 
 
+class ChunkedDistributedLogprob(torch.autograd.Function):
```
Comment (terrykong) [inline nemo_rl/distributed/model_utils.py:143]: can we add a unit test for this function so we make sure the non-chunked version equals the chunk?

File: nemo_rl/tron/model.py (hunk 4 above)
Comment (terrykong) [reply nemo_rl/tron/model.py]: will check internally

File: nemo_rl/models/policy/megatron_policy_worker.py [hunk 7]
```diff
@@ -1099,6 +1105,8 @@ def collection_fn(output_tensor):
                         group=tp_grp,
                         inference_only=True,
                         cp_group=get_context_parallel_group(),
+                        # TODO(pjin): chunked logprob not implemented yet w/ seq packing.
```
Comment (wangshangsam) [inline nemo_rl/models/policy/megatron_policy_worker.py]: @pjin-nvidia Could you file an issue for this? `#TODO` has a tendency of being lost in the code.

File: nemo_rl/tron/model.py (hunk 5 above)
Comment (pjin-nvidia) [reply nemo_rl/tron/model.py]: it's a copy of nemo/tron/model.py:
https://github.com/NVIDIA/NeMo/blob/8ddf4387344c6423763ec9ee0c9a755cbb5d8d35/nemo/tron/model.py

the main change is removing the `Float16Module` wrapper (which is what originally casts the model logits output to float32):
https://github.com/NVIDIA-NeMo/RL/pull/856/files/a020289609cfa0d7a695a175eed009fdb4695088#diff-37539801eab6c58172c5cf85be33a1f9eac04c096a8e23170550ddf3bff8e3b3R125-R128

File: nemo_rl/models/policy/megatron_policy_worker.py (hunk 7 above)
Comment (pjin-nvidia) [reply nemo_rl/models/policy/megatron_policy_worker.py]: I ended up just adding optional `chunk_size` to `from_parallel_logits_to_logprobs_packed_sequences` too

File: nemo_rl/distributed/model_utils.py (hunk 1 above)
Comment (pjin-nvidia) [reply nemo_rl/distributed/model_utils.py:266]: added

File: nemo_rl/distributed/model_utils.py (hunk 2 above)
Comment (pjin-nvidia) [reply nemo_rl/distributed/model_utils.py:351]: added

File: nemo_rl/models/dtensor/parallelize.py (hunk 3 above)
Comment (pjin-nvidia) [reply nemo_rl/models/dtensor/parallelize.py:699]: added

File: nemo_rl/distributed/model_utils.py (hunk 6 above)
Comment (pjin-nvidia) [reply nemo_rl/distributed/model_utils.py:143]: added a `chunk_size` parameter to `DistributedLogprobTestActor`:

https://github.com/NVIDIA-NeMo/RL/pull/856/files#diff-13516af2fb4ffe4b66dd0cc8a1113bcd8334a1500f77ab0ce0f503bf529d8dc9

File: examples/configs/grpo_math_qwen3_30ba3b_megatron_tp4_32k.yaml [hunk 8]
```diff
@@ -0,0 +1,103 @@
+# GRPO Algorithm Configuration
+defaults: "grpo_math_1B_megatron.yaml"
```
Comment (terrykong) [inline examples/configs/grpo_math_qwen3_30ba3b_megatron_tp4_32k.yaml]: can you put this config into `examples/configs/recipes/llm`? the top level configs are intended as exemplars and are subject to change, whereas the ones in recipes are "static"

also adding it under `./recipes` will cause a unit test to fail if you don't have a corresponding functional test for this under https://github.com/NVIDIA-NeMo/RL/tree/main/tests/test_suites/llm, so please run the unit tests to make sure you've added everything necessary

File: nemo_rl/tron/model.py (hunk 5 above)
Comment (terrykong) [reply nemo_rl/tron/model.py]: If it's only a one-line change, I'd prefer the change be reflected in the submodule (you can branch where the submodule is at to update)

also, if you expect the model coming back to not be a FP16 but something else, could you add a test asserting the model type? We're currently migrating away from tron, so once that is done, this test would ensure we don't miss this typing fix you're adding

Comment (pjin-nvidia) [reply nemo_rl/tron/model.py]: updated NeMo submodule

branch: https://github.com/NVIDIA/NeMo/tree/pjin/nemorl-logprob
commit: https://github.com/NVIDIA/NeMo/commit/0bf0dbce6c6794efc0df4d0980773c14d9632421

Comment (pjin-nvidia) [reply nemo_rl/tron/model.py]: > also, if you expect the model coming back to not be a FP16 but something else, could you add a test asserting the model type? We're currently migrating away from tron, so once that is done, this test would ensure we don't miss this typing fix you're adding

what I did is add a float32 dtype check to the existing megatron logprobs test, and running that test on more cases of `(logprob chunk size, deferred float32 logits)`
https://github.com/NVIDIA-NeMo/RL/pull/856/files#diff-9556cb57e37308923c54e7a6df8982afafef5e36544f350af3324db43f74bdbeR703

one thing is that the policy model worker mainly exposes the model output through `get_logprobs`, and there is not another interface for getting at the underlying torch model logits. but I think just checking that the returned logprobs are float32 should be sufficient?

File: nemo_rl/models/policy/megatron_policy_worker.py (hunk 7 above)
Comment (wangshangsam) [reply nemo_rl/models/policy/megatron_policy_worker.py]: Sounds good. Thanks!

File: examples/configs/grpo_math_qwen3_30ba3b_megatron_tp4_32k.yaml (hunk 8 above)
Comment (pjin-nvidia) [reply examples/configs/grpo_math_qwen3_30ba3b_megatron_tp4_32k.yaml]: adding this test training script:
https://github.com/NVIDIA-NeMo/RL/blob/pjin/logprob/tests/test_suites/llm/grpo-math-qwen3-30ba3b-megatron-tp4-32k.sh

test run results:
https://wandb.ai/nvidia/nemo-rl/runs/h74m92l7/

File: nemo_rl/distributed/model_utils.py [hunk 9]
```diff
@@ -141,6 +143,123 @@ def backward(
...
+            chunk_end = min(seq_size, (chunk_idx + 1) * chunk_size)
+
+            logits = vocab_parallel_logits[:, chunk_start:chunk_end, :]
+            logits = logits.to(dtype=torch.float32)
+
+            log_softmax_output = _compute_distributed_log_softmax(
+                logits,
+                group=tp_group,
+            )
+            log_probs = log_softmax_output.clone()
```
Comment (soodoshll) [inline nemo_rl/distributed/model_utils.py]: qq - what is the purpose of making a copy here?

File: nemo_rl/models/policy/dtensor_policy_worker.py [hunk 10]
```diff
@@ -1056,7 +1058,7 @@ def get_logprobs(
                             #   token_ids: [batch_size, sequence_length] - actual tokens
                             # Output shape: [batch_size, sequence_length] - logprob of each token given previous
                             # We get logprob of token[t+1] from logits[t], prepending 0 to maintain sequence length
-                            logits = outputs.logits.to(torch.float32)
+                            logits = logits.to(torch.float32)
```
Comment (soodoshll) [inline nemo_rl/models/policy/dtensor_policy_worker.py:1075]: wondering if chunking also helps with the non-distributed path?

Comment (pjin-nvidia) [reply nemo_rl/models/policy/dtensor_policy_worker.py:1075]: yeah good point, I added a chunked path to the non-distributed case too

File: examples/configs/recipes/llm/grpo-math-qwen3-30ba3b-megatron-tp4-32k.yaml [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+    pipeline_dtype: ${policy.precision}
+    num_layers_in_first_pipeline_stage: null
+    num_layers_in_last_pipeline_stage: null
+    freeze_moe_router: True
+    moe_router_dtype: "fp64"
+    moe_router_load_balancing_type: "none" # "seq_aux_loss" causes logprob error divergence for grpo
+    moe_router_bias_update_rate: 0.0 # by default, disable bias updates for grpo
+    apply_rope_fusion: True
+    activation_checkpointing: True
+    defer_fp32_logits: True
```
Comment (terrykong) [inline examples/configs/recipes/llm/grpo-math-qwen3-30ba3b-megatron-tp4-32k.yaml:81]: what would be the reason to set this to False?

File: nemo_rl/tron/model.py (hunk 4 above)
Comment (terrykong) [reply nemo_rl/tron/model.py]: guidance from internal was to leave this as is

File: nemo_rl/distributed/model_utils.py (hunk 9 above)
Comment (pjin-nvidia) [reply nemo_rl/distributed/model_utils.py]: removed unnecessary clone

File: examples/configs/recipes/llm/grpo-math-qwen3-30ba3b-megatron-tp4-32k.yaml (hunk 11 above)
Comment (pjin-nvidia) [reply examples/configs/recipes/llm/grpo-math-qwen3-30ba3b-megatron-tp4-32k.yaml:81]: mostly for strict backward compat, but we could instead enable it by default (i.e. make it an opt-out config like `no_defer_fp32_logits` or similar)

wdyt?

Comment (terrykong) [reply examples/configs/recipes/llm/grpo-math-qwen3-30ba3b-megatron-tp4-32k.yaml:81]: I see. How about the following:
1. this PR introduces it, default off
2. follow up PR where we run all our nightly tests to see if defaulting to true is ok, if so, remove the arg
wdyt? If the feature is broadly applicable we should probably switch it to true so no one else runs into the same issue (assuming no accuracy penalty)

File: nemo_rl/distributed/model_utils.py (hunk 9 above)
Comment (soodoshll) [reply nemo_rl/distributed/model_utils.py]: there is another one in the backward?

File: examples/configs/recipes/llm/grpo-math-qwen3-30ba3b-megatron-tp4-32k.yaml (hunk 11 above)
Comment (pjin-nvidia) [reply examples/configs/recipes/llm/grpo-math-qwen3-30ba3b-megatron-tp4-32k.yaml:81]: yup, (1) and then (2) SGTM!

File: nemo_rl/distributed/model_utils.py (hunk 9 above)
Comment (pjin-nvidia) [reply nemo_rl/distributed/model_utils.py]: I replaced a couple remaining instances of `clone()` followed by `exp_()` with just `exp()` and no clone

**Comment by terrykong** [comment]
There's a permission issue with that `Check submodule fast-forward` job. @chtruong814 is taking a look

**Comment by pjin-nvidia** [comment]
closing in favor of https://github.com/NVIDIA-NeMo/RL/pull/918
